### PR TITLE
Fix #14958 - remove sales sequence data on store view delete

### LINKED ISSUE #14958: sale_sequence_* records are not removed on store view deletion
`sale_sequence_meta` and `sales_sequence_profile` records are not removed upon store deletion.

### Preconditions
1. Magento 2.2

### Steps to reproduce
1. Delete existing store view

### Expected result
1. `sale_sequence_meta` and `sales_sequence_profile` contain no records referencing to removed store

### Actual result
1. `sale_sequence_meta` and `sales_sequence_profile` do contain records referencing to removed store

### PR DESCRIPTION
**Pull request by Bartlomiejsz**
In this PR I tried to fix #18466 after it was probably abaddoned.

### Fixed Issues (if relevant)

1. magento/magento2#14958: sale_sequence_* records are not removed on store view deletion

### Manual testing scenarios (*)

1. Delete existing store view
2. Check **sale_sequence_meta** and **sales_sequence_profile** tables - those shouldn't contain any data related to deleted store view.
3. All sequence tables for deleted store view should not exist.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/SalesSequence/Model/Builder.php [hunk 1]
```diff
@@ -264,4 +274,36 @@ public function create()
...
+
+    /**
+     * Deletes all sequence linked entites
+     *
+     * @param $storeId
+     *
+     * @return void
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function deleteByStoreId($storeId)
```
Comment (dmytro-ch) [inline app/code/Magento/SalesSequence/Model/Builder.php]: Hi @Bartlomiejsz!
Looks like the `Builder` class is not the proper place for this `delete` method.
Could you please move this logic into a new separate class `Magento\SalesSequence\Model\Sequence\DeleteByStore` that provides only one public method `execute(StoreInterface $store)`.

Please check the `\Magento\CustomerGraphQl\Model\Customer\ChangeSubscriptionStatus` as an example.

Thank you!

**Review by dmytro-ch — CHANGES_REQUESTED** [review]

File: app/code/Magento/SalesSequence/Model/Builder.php (hunk 1 above)
Comment (Bartlomiejsz) [reply app/code/Magento/SalesSequence/Model/Builder.php]: Hi @dmytro-ch, done :)

**Review by dmytro-ch — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch, thank you for the review.
[ENGCOM-4781](https://jira.corp.magento.com/browse/ENGCOM-4781) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Comment by nmalevanec** [comment]
@sivaschenko, there is some backward incompatible changes, made in this pull request. Could you please take a look?
1. Public method getIdsByStore() has been added into api class Magento\SalesSequence\Model\ResourceModel\Meta
2. Public method getProfileIdsByMetadataIds has been added onto api class Magento\SalesSequence\Model\ResourceModel\Profile
3. New api class Magento\SalesSequence\Model\Sequence\DeleteByStore has been introduced.

File: app/code/Magento/SalesSequence/Model/Sequence/DeleteByStore.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+use Magento\Framework\App\ResourceConnection as AppResource;
+use Magento\SalesSequence\Model\MetaFactory;
+use Magento\SalesSequence\Model\ResourceModel\Meta as ResourceMetadata;
+use Magento\SalesSequence\Model\ResourceModel\Profile as ResourceProfile;
+use Magento\Store\Api\Data\StoreInterface;
+
+/**
+ * Class DeleteByStore
+ * @api
```
Comment (sivaschenko) [inline app/code/Magento/SalesSequence/Model/Sequence/DeleteByStore.php]: Please remove the `@api` annotation, API cannot be introduced to patch releases

File: app/code/Magento/SalesSequence/Model/ResourceModel/Meta.php [hunk 3]
```diff
@@ -92,6 +92,27 @@ public function loadByEntityTypeAndStore($entityType, $storeId)
...
     }
 
+    /**
+     * Retrieves Metadata Ids by store id
+     *
+     * @param int $storeId
+     * @return int[]
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function getIdsByStore($storeId)
```
Comment (sivaschenko) [inline app/code/Magento/SalesSequence/Model/ResourceModel/Meta.php]: Introduction of a new public method is backwards incompatible and cannot be added to a patch release. Please create a new class for this functionality

File: app/code/Magento/SalesSequence/Model/ResourceModel/Profile.php [hunk 4]
```diff
@@ -76,4 +77,21 @@ public function loadActiveProfile($metadataId)
...
     }
+
+    /**
+     * Get profile ids by metadata ids
+     *
+     * @param int[] $metadataIds
+     * @return int[]
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function getProfileIdsByMetadataIds(array $metadataIds)
```
Comment (sivaschenko) [inline app/code/Magento/SalesSequence/Model/ResourceModel/Profile.php]: Introduction of a new public method is backwards incompatible and cannot be added to a patch release. Please create a new class for this functionality

File: app/code/Magento/SalesSequence/composer.json [hunk 5]
```diff
@@ -6,7 +6,8 @@
     },
     "require": {
         "php": "~7.1.3||~7.2.0",
-        "magento/framework": "*"
+        "magento/framework": "*",
+        "magento/module-store": "*"
```
Comment (sivaschenko) [inline app/code/Magento/SalesSequence/composer.json]: @akaplya what do you think about adding this dependency?

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the pull request @Bartlomiejsz ! There are several backward incompatible changes that have to be resolved.

For more details on backward compatible development see: https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

**Comment by Bartlomiejsz** [comment]
Hi @sivaschenko, all changes implemented!

**Comment by sivaschenko** [comment]
Hi @Bartlomiejsz thanks for your updates!

After the refactoring, it is clear that the two added `Ids` classes and their methods are used only once in the introduced `DeleteByStore` model. Are the `Ids` classes intended to be reused or could the functionality of loading meta by stores and profiles by meta be simply included as a private functionality of the introduced `DeleteByStore` model?

**Comment by Bartlomiejsz** [comment]
Hi @sivaschenko, it's hard to say if those methods will be reused. In current state I believe not. And of course those can be moved to private methods if required.

**Comment by Bartlomiejsz** [comment]
Hi @sivaschenko, I removed additional dependency on store module, this is no longer needed. Please verify now :)

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4781](https://jira.corp.magento.com/browse/ENGCOM-4781) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by sidolov** [comment]
@engcom-Foxtrot could you please take a look at failed integration build? 
The issue reproducing only on the Magento Commerce and B2B edition, fatal error in logs is present: 
```
{"messages":{"error":[{"code":500,"message":"Fatal Error: 'Uncaught Magento\\Framework\\DB\\Adapter\\ConnectionException: DDL statements are not allowed in transactions in \/var\/www\/html\/lib\/internal\/Magento\/Framework\/DB\/Adapter\/Pdo\/Mysql.php:515\nStack trace:\n#0 \/var\/www\/html\/lib\/internal\/Magento\/Framework\/DB\/Adapter\/Pdo\/Mysql.php(543): Magento\\Framework\\DB\\Adapter\\Pdo\\Mysql->_checkDdlTransaction('CREATE TABLE se...')\n#1 \/var\/www\/html\/lib\/internal\/Magento\/Framework\/DB\/Adapter\/Pdo\/Mysql.php(612): Magento\\Framework\\DB\\Adapter\\Pdo\\Mysql->_query('CREATE TABLE se...', Array)\n#2 \/var\/www\/html\/dev\/tests\/integration\/tmp\/sandbox--76ccf56fc1cd540134f59b4069e60caaa4132cb3976271c3398ab8e1c4c45409\/generated\/code\/Magento\/TestFramework\/Db\/Adapter\/Mysql\/Interceptor.php(167): Magento\\Framework\\DB\\Adapter\\Pdo\\Mysql->query('CREATE TABLE se...', Array)\n#3 \/var\/www\/html\/app\/code\/Magento\/ResourceConnections\/DB\/Adapter\/Pdo\/MysqlProxy.php(266): Magento\\TestFramework\\Db\\Adapter\\Mysql\\Interceptor->query('CREATE TABLE se...', Array)\n#4 \/var\/www\/html\/dev\/tests\/integration\/tmp\/sandbox--76ccf56fc1cd54' in '\/var\/www\/html\/lib\/internal\/Magento\/Framework\/DB\/Adapter\/Pdo\/Mysql.php' on line 515","trace":"Trace is not available."}]}}
```

Looks like some tests in Commerce or B2B edition uses fixtures with tests isolation and failing due to DDL operation introduced in current PR.

**Comment by Bartlomiejsz** [comment]
Hi @engcom-Foxtrot, any news regarding this?

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Pull request merged**